### PR TITLE
Add option to collect identifiers from tags file

### PR DESCRIPTION
**Pull request by ptrv**

### CODE REVIEW

File: ycmd.el [hunk 1]
```diff
@@ -352,6 +357,23 @@ Returns the new value of `ycmd-force-semantic-completion`.
              (if force "enabled" "disabled"))
     (setq ycmd-force-semantic-completion force)))
 
+(defun ycmd--get-tags-file (buffer)
+  "Get tags files for current BUFFER or nil.
+
+Use `tags-file-name' variable if set, or look up directory
+hierarchy for first found tags file."
+  (or tags-file-name
+      (let* ((directory (locate-dominating-file
```
Comment (ptrv) [inline ycmd.el]: Should we do this? Going up the directory hierarchy for a possible tags file? or should we only use `tags-file-name` or should we event introduce a variable in ycmd?

Comment (abingham) [inline ycmd.el]: It seems like we might need a combination of the two. I think the `locate-dominating-file` approach is good for a lot of people, but they might need to search for files named something other than `tags` or 'TAGS`, e.g. if they use `CTAGS` as their tag file name by default. But the `tags-file-name` idea will be useful if people use shared tag files. 

Honestly, I don't know enough about how people use tag files to be positive. I think we should create a system that seems reasonable, and which addresses the needs of a known user. We can polish it as we get input.

**Comment by abingham** [comment]
@henry-y Can you try this out and let us know if it works for you?

**Comment by ptrv** [comment]
@abingham Here is a full-fledged solution with different options for tag files locations.

**Comment by henry-y** [comment]
guys, I don't know why it won't work for me.
from ycmd-content-log output, it seems it goes the following steps:

1. ask me if I want to load .ycm_extra_conf.py? so I answered yes, the related log is:

[omitted logs.......]
sForFile\n    module = extra_conf_store.ModuleForSourceFile( filename )\n  File \"/root/dev/ycmd/ycmd/../ycmd/extra_conf_store.py\", line 48, in ModuleForSourceFile\n    retu\
rn Load( ModuleFileForSourceFile( filename ) )\n  File \"/root/dev/ycmd/ycmd/../ycmd/extra_conf_store.py\", line 59, in ModuleFileForSourceFile\n    if Load( module_file ):\n\
  File \"/root/dev/ycmd/ycmd/../ycmd/extra_conf_store.py\", line 134, in Load\n    if not _ShouldLoad( module_file ):\n  File \"/root/dev/ycmd/ycmd/../ycmd/extra_conf_store.p\
y\", line 117, in _ShouldLoad\n    raise UnknownExtraConf( module_file )\nUnknownExtraConf: Found /root/dev/mc/.ycm_extra_conf.py. Load? \n\n(Question can be turned off with \
options, see YCM docs)\n")                                                                                                                                                     
 (exception                                                                                                                                                                    
  (extra_conf_file . "/root/dev/mc/.ycm_extra_conf.py")                                                                                                                        
  (TYPE . "UnknownExtraConf")))                                                                                                                                                
                                                                                                                                                                               
HTTP REQUEST CONTENT                                                                                                                                                           
                                                                                                                                                                               
"{\"filepath\":\"\\/root\\/dev\\/mc\\/.ycm_extra_conf.py\"}"                                                                                                                   
HTTP RESPONSE CONTENT                                                                                                                                                          
                                                                                                                                                                               
""                                                                                                                                                                             
HTTP REQUEST CONTENT                                                                                                                                                           
                                                                                                                                                                               
"{\"event_name\":\"FileReadyToParse\",\"file_data\":{\"\\/root\\/dev\\/mc\\/src\\/mc_client\\/src\\/mc_client_logic.c\":{\"contents\":\"#include \\\"mc_client_logic.h\\\"\\n#\
include \\\"mc_log.h\\\"\\n#include \\\"lib_ch.h\\\"\\n#includ
[omitted logs.......]
delete the downlink.\\\\n\\\");\\n        return RETURN_ERROR;\\n    }\\n\\n    \\/\\/LOG_DBG(\\\"[uninitialize] succ.\\\\n\\\");\\n    return RETURN_OK;\\n}\\n\\n\\n\\n\
\",\"filetypes\":[\"cpp\"]}},\"filepath\":\"\\/root\\/dev\\/mc\\/src\\/mc_client\\/src\\/mc_client_logic.c\",\"line_num\":19,\"column_num\":1,\"tag_files\":[\"\\/root\\/dev\\\
/tags\"]}"                                                                                                                                                                     
HTTP RESPONSE CONTENT                                                                                                                                                          
                                                                                                                                                                               
[((location_extent                                                                                                                                                             
   (end                                                                                                                                                                        
    (line_num . 18)                                                                                                                                                            
    (column_num . 42)                                                                                                                                                          
    (filepath . "/root/dev/mc/src/lib_msg/include/lib_msg.h"))                                                                                                                 
   (start                                                                                                                                                                      
    (line_num . 18)                                                                                                                                                            
    (column_num . 20)                                                                                                                                                          
    (filepath . "/root/dev/mc/src/lib_msg/include/lib_msg.h")))                                                                                                                
  (location                                                                                                                                                                    
   (line_num . 18)                                                                                                                                                             
   (column_num . 20)                                                                                                                                                           
   (filepath . "/root/dev/mc/src/lib_msg/include/lib_msg.h"))                                                                                                                  
  (kind . "ERROR")                                                                                                                                                             
  (text . "use of undeclared identifier 'NAME_MAX'")                                                                                                                           
  (ranges .                                                                                                                                                                    
          []))                                                                                                                                                                 
 ((location_extent                                    



so I can see there's something about tag file in the request content above:
"tag_files\":[\"\\/root\\/dev\\\
/tags\"]}"

2. then I typed some text, but there won't be any tag_files in the request anymore, log is like this:

\\n        LOG_ERR(\\\"It's failed to unregister to server.\\\\n\\\");\\n        return RETURN_ERROR;\\n    }\\n\\n    return RETURN_OK;\\n}\\n\\nreturn_t MCCLIENT_uninitiali\
ze(void)\\n{\\n    s32 rlt = 0;\\n    \\n    rlt = LIBCH_close(local_fd);\\n    if(-1 == rlt)\\n    {\\n        LOG_ERR(\\\"It's failed to close the downlink.\\\\n\\\");\\n  \
      return RETURN_ERROR;\\n    }\\n\\n    \\n    rlt = LIBCH_delete(client_name);\\n    if(-1 == rlt)\\n    {\\n        LOG_ERR(\\\"It's failed to delete the downlink.\\\\n\
\\\");\\n        return RETURN_ERROR;\\n    }\\n\\n    \\/\\/LOG_DBG(\\\"[uninitialize] succ.\\\\n\\\");\\n    return RETURN_OK;\\n}\\n\\n\\n\\n\",\"filetypes\":[\"cpp\"]}},\\
"filepath\":\"\\/root\\/dev\\/mc\\/src\\/mc_client\\/src\\/mc_client_logic.c\",\"line_num\":19,\"column_num\":7}"                                                              
HTTP RESPONSE CONTENT                                                                                                                                                          
                                                                                                                                                                               
((completion_start_column . 5)                                                                                                                                                 
 (completions .                                                                                                                                                                
              [((insertion_text . "MCCLIENT_read_config"))                                                                                                                     
               ((insertion_text . "MAX_CH_BUF_SIZE"))                                                                                                                          
               ((insertion_text . "MAX_CLIENT_NAME_LENGTH"))                                                                                                                   
               ((insertion_text . "MC_CONFIG_FILE_NAME_PATH"))                                                                                                                 
               ((insertion_text . "MCCLIENT_setname"))                                                                                                                         
               ((insertion_text . "MCCLIENT_register"))                                                                                                                        
               ((insertion_text . "MCCLIENT_initialize"))                                                                                                                      
               ((insertion_text . "MCCLIENT_unregister"))                                                                                                                      
               ((insertion_text . "MCCLIENT_sendmessage"))                                                                                                                     
               ((insertion_text . "MCCLIENT_uninitialize"))]))                                                                                                                 
                                                                                                                                                                               
HTTP REQUEST CONTENT                                                                                                                                                           
                                                                                                                                                                               
"{\"file_data\":{\"\\/root\\/dev\\/mc\\/src\\/mc_client\\/src\\/mc_client_logic.c\":{\"contents\":\"#include \\\"mc_client_logic.h\\\"\\n#include \\\"mc_log.h\\\"\\n#include \
\\\"lib_ch.h\\\"\\n#include \\\"lib_msg.h\\\"\\n#include <sys\\/types.h>\\n#include <sys\\/time.h>\\n#include <sys\\/stat.h>\\n#include <sys\\/select.h>\\n\\n#define MAX_SELE\
CT_TIME_OUT                     1\\n\\nstatic s32 remote_fd = 0;\



you can see from above that the end line of the request content doesn't contain tag_files.

**Comment by abingham** [comment]
I *think* you should not expect to see tag-file information in the last request. You should only expect to see it in the `FileReadyToParse` request, and that seems to be happening in the log you showed. So it looks like we're sending `ycmd` the information it needs, but you're still not seeing the completions you expect.

Have you ensured that you're following [all of the rules for tag files](https://github.com/Valloric/YouCompleteMe#the-gycm_collect_identifiers_from_tags_files-option)? You seem to be doing everything correctly, so I'm grasping at straws a bit here.

**Comment by henry-y** [comment]
@abingham Thanks, I forgot the rules for tag files, it works for me now.

but since my tags file is pretty large (946M), it's very slow for the completions candidates to pop up. Is there any way to improve?

**Comment by abingham** [comment]
With a tag file of that size, we may be a bit stuck. From the point of view of `emacs-ycmd`, we just send a request to `ycmd` and wait for an answer. Perhaps there's some way to make multiple requests, i.e. one which doesn't use tags and one which *only* uses tags.

@valloric Do you have any suggestions for working with such large tag files?

**Comment by Valloric** [comment]
@abingham My suggestion would be "don't use such large tag files." :)

Seriously though, the tag file listed is a gig of text. Note that an N-byte identifier ends up using N*X bytes of RAM (where X is something around 20 probably, if not more) in ycmd because each character becomes a node in a complex trie-like structure. This is to facilitate fast matching. You might be hitting swap with such a large tag file.

**Pull request merged**